File: src/realforge/provider_image_gen.py
"""Approval-gated local image generation (sandbox-only).

One bounded prompt -> the user-configured local image backend -> one base64 PNG.
Two pluggable backends:
  * ``local_image_provider`` — any OpenAI-compatible ``/images/generations`` server.
  * ``comfyui`` — a local ComfyUI install (queue ``/prompt``, poll ``/history``,
    fetch ``/view``); the user's API-format workflow carries a ``%prompt%`` token.
No model name, API key, path, workflow, or private config is ever returned. Output
is LOCAL UNTRUSTED and not persisted here.
"""

from __future__ import annotations

import base64
import binascii
import json
import time
import unicodedata
from collections.abc import Callable
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any
from urllib.parse import urlencode, urljoin
from uuid import uuid4

from realforge.private_provider_config import (
    COMFYUI_IMAGE_PROVIDER,
    LOCAL_IMAGE_PROVIDER,
    PrivateImageProviderRuntimeSettings,
    PrivateProviderConfigError,
    load_private_local_config_bundle,
    parse_local_endpoint,
)
from realforge.providers.http_util import HTTPProviderError, get_bytes, get_json, post_json

IMAGE_MAX_PROMPT_CHARS = 2_000
IMAGE_TIMEOUT_SECONDS = 60.0  # ponytail: image gen is slow; generous vs chat's 20s
IMAGE_MAX_BYTES = 8 * 1024 * 1024  # decoded PNG cap
IMAGE_MAX_RESPONSE_BYTES = 12 * 1024 * 1024  # raw JSON cap (b64 is ~33% larger)
COMFYUI_PROMPT_TOKEN = "%prompt%"  # injected into the user's workflow text node
COMFYUI_POLL_INTERVAL_SECONDS = 0.5
COMFYUI_MAX_WORKFLOW_BYTES = 1024 * 1024  # workflow_path file cap
PNG_SIGNATURE = b"\x89PNG\r\n\x1a\n"


@dataclass(frozen=True)
class ProviderImageError:
    code: str
    message: str


@dataclass(frozen=True)
class ProviderImageReport:
    ok: bool
    attempted: bool
    configured: bool
    status: str  # pass | not_configured | rejected | fail
    input_length: int
    duration_ms: int
    image_base64: str | None
    mime: str | None
    image_bytes: int
    untrusted_output: bool
    error: ProviderImageError | None

    def to_dict(self) -> dict[str, object]:
        return asdict(self)


def run_private_provider_image_gen(
    prompt: str,
    *,
    opener: Callable[..., Any] | None = None,
    clock: Callable[[], float] = time.perf_counter,
    sleep: Callable[[float], None] = time.sleep,
) -> ProviderImageReport:
    started = clock()
    validated, err = _validate_prompt(prompt)
    if err is not None:
        return _fail(started, clock, err.code, err.message, status="rejected")
    assert validated is not None
    input_length = len(validated)

    try:
        bundle = load_private_local_config_bundle()
    except PrivateProviderConfigError:
        return _fail(started, clock, "invalid_config", "Private local provider configuration is invalid.", input_length=input_length)

    image = bundle.image
    if image is None or not image.configured:
        return _fail(started, clock, "not_configured", "Private local image provider is not configured.", status="not_configured", input_length=input_length)
    if parse_local_endpoint(image.base_url or "") is None:
        return _fail(started, clock, "not_configured", "Private local image provider endpoint is missing or not local.", status="not_configured", input_length=input_length, configured=True)

    if image.kind == LOCAL_IMAGE_PROVIDER:
        return _generate_openai_compatible(image, validated, input_length, started, clock, opener)
    if image.kind == COMFYUI_IMAGE_PROVIDER:
        return _generate_comfyui(image, validated, input_length, started, clock, opener, sleep)
    return _fail(started, clock, "not_configured", "Private local image provider backend is not supported.", status="not_configured", input_length=input_length, configured=True)


def format_provider_image_gen_json(report: ProviderImageReport) -> str:
    return json.dumps(report.to_dict(), sort_keys=True)


# --- OpenAI-compatible /images/generations backend -------------------------------


def _generate_openai_compatible(
    image: PrivateImageProviderRuntimeSettings,
    validated: str,
    input_length: int,
    started: float,
    clock: Callable[[], float],
    opener: Callable[..., Any] | None,
) -> ProviderImageReport:
    url = urljoin((image.base_url or "") + "/", "images/generations")
    payload: dict[str, Any] = {"prompt": validated, "n": 1, "response_format": "b64_json"}
    if image.model:
        payload["model"] = image.model
    headers = {"Authorization": f"Bearer {image.api_key}"} if image.api_key else {}

    try:
        data = post_json(url, payload, timeout=IMAGE_TIMEOUT_SECONDS, extra_headers=headers, opener=opener, max_response_bytes=IMAGE_MAX_RESPONSE_BYTES)
    except HTTPProviderError as exc:
        return _fail(started, clock, exc.code, _redact(exc.code), attempted=True, input_length=input_length, configured=True)
    except Exception:  # noqa: BLE001 - never surface provider internals or private values
        return _fail(started, clock, "provider_error", "Local image provider request failed.", attempted=True, input_length=input_length, configured=True)

    b64, decode_error = _extract_image(data)
    if decode_error is not None:
        return _fail(started, clock, decode_error.code, decode_error.message, attempted=True, input_length=input_length, configured=True)
    assert b64 is not None
    return _ok(started, clock, input_length, b64, len(base64.b64decode(b64)))


def _extract_image(data: dict) -> tuple[str | None, ProviderImageError | None]:
    items = data.get("data")
    first = items[0] if isinstance(items, list) and items and isinstance(items[0], dict) else None
    b64 = first.get("b64_json") if isinstance(first, dict) else None
    if not isinstance(b64, str) or not b64:
        return None, ProviderImageError("invalid_response", "Local image provider returned no image data.")
    try:
        raw = base64.b64decode(b64, validate=True)
    except (binascii.Error, ValueError):
        return None, ProviderImageError("invalid_response", "Local image provider returned malformed image data.")
    error = _validate_png(raw)
    if error is not None:
        return None, error
    return b64, None


# --- ComfyUI backend -------------------------------------------------------------


def _generate_comfyui(
    image: PrivateImageProviderRuntimeSettings,
    validated: str,
    input_length: int,
    started: float,
    clock: Callable[[], float],
    opener: Callable[..., Any] | None,
    sleep: Callable[[float], None],
) -> ProviderImageReport:
    workflow_text, load_err = _load_comfyui_workflow(image)
    if load_err is not None:
        return _fail(started, clock, load_err.code, load_err.message, input_length=input_length, configured=True)
    assert workflow_text is not None

    injected, inject_err = _inject_prompt(workflow_text, validated)
    if inject_err is not None:
        return _fail(started, clock, inject_err.code, inject_err.message, input_length=input_length, configured=True)
    try:
        workflow = json.loads(injected)
    except json.JSONDecodeError:
        return _fail(started, clock, "invalid_workflow", "ComfyUI workflow is not valid JSON after prompt injection.", input_length=input_length, configured=True)
    if not isinstance(workflow, dict):
        return _fail(started, clock, "invalid_workflow", "ComfyUI workflow must be a node graph object.", input_length=input_length, configured=True)

    base = (image.base_url or "").rstrip("/")
    try:
        queued = post_json(urljoin(base + "/", "prompt"), {"prompt": workflow, "client_id": uuid4().hex}, timeout=IMAGE_TIMEOUT_SECONDS, opener=opener, max_response_bytes=IMAGE_MAX_RESPONSE_BYTES)
    except HTTPProviderError as exc:
        return _fail(started, clock, exc.code, _redact(exc.code), attempted=True, input_length=input_length, configured=True)
    except Exception:  # noqa: BLE001 - never surface provider internals or private values
        return _fail(started, clock, "provider_error", "Local image provider request failed.", attempted=True, input_length=input_length, configured=True)

    prompt_id = queued.get("prompt_id")
    if not isinstance(prompt_id, str) or not prompt_id:
        return _fail(started, clock, "invalid_response", "ComfyUI did not accept the workflow.", attempted=True, input_length=input_length, configured=True)

    ref, poll_err = _poll_comfyui_history(base, prompt_id, opener, sleep, started, clock)
    if poll_err is not None:
        return _fail(started, clock, poll_err.code, poll_err.message, attempted=True, input_length=input_length, configured=True)
    assert ref is not None

    raw, fetch_err = _fetch_comfyui_image(base, ref, opener)
    if fetch_err is not None:
        return _fail(started, clock, fetch_err.code, fetch_err.message, attempted=True, input_length=input_length, configured=True)
    assert raw is not None

    png_err = _validate_png(raw)
    if png_err is not None:
        return _fail(started, clock, png_err.code, png_err.message, attempted=True, input_length=input_length, configured=True)
    return _ok(started, clock, input_length, base64.b64encode(raw).decode("ascii"), len(raw))


def _load_comfyui_workflow(image: PrivateImageProviderRuntimeSettings) -> tuple[str | None, ProviderImageError | None]:
    if image.workflow:
        return image.workflow, None
    if image.workflow_path:
        try:
            path = Path(image.workflow_path).expanduser()
            if not path.is_file():
                return None, ProviderImageError("workflow_missing", "Configured ComfyUI workflow file was not found.")
            if path.stat().st_size > COMFYUI_MAX_WORKFLOW_BYTES:
                return None, ProviderImageError("workflow_too_large", "Configured ComfyUI workflow file is too large.")
            return path.read_text(encoding="utf-8"), None
        except OSError:
            return None, ProviderImageError("workflow_missing", "Configured ComfyUI workflow file could not be read.")
    return None, ProviderImageError("not_configured", "ComfyUI workflow is not configured.")


def _inject_prompt(workflow_text: str, prompt: str) -> tuple[str | None, ProviderImageError | None]:
    if COMFYUI_PROMPT_TOKEN not in workflow_text:
        return None, ProviderImageError("workflow_no_placeholder", f"ComfyUI workflow must contain the {COMFYUI_PROMPT_TOKEN} token in a text node.")
    # json.dumps escapes the prompt for embedding inside an existing JSON string.
    escaped = json.dumps(prompt)[1:-1]
    return workflow_text.replace(COMFYUI_PROMPT_TOKEN, escaped), None


def _poll_comfyui_history(
    base: str,
    prompt_id: str,
    opener: Callable[..., Any] | None,
    sleep: Callable[[float], None],
    started: float,
    clock: Callable[[], float],
) -> tuple[dict[str, str] | None, ProviderImageError | None]:
    url = urljoin(base + "/", f"history/{prompt_id}")
    while True:
        try:
            history = get_json(url, timeout=IMAGE_TIMEOUT_SECONDS, opener=opener, max_response_bytes=IMAGE_MAX_RESPONSE_BYTES)
        except HTTPProviderError as exc:
            return None, ProviderImageError(exc.code, _redact(exc.code))
        except Exception:  # noqa: BLE001 - never surface provider internals
            return None, ProviderImageError("provider_error", "Local image provider request failed.")
        entry = history.get(prompt_id) if isinstance(history, dict) else None
        if isinstance(entry, dict):
            ref = _first_image_ref(entry.get("outputs"))
            if ref is not None:
                return ref, None
            status = entry.get("status")
            if isinstance(status, dict) and status.get("status_str") == "error":
                return None, ProviderImageError("provider_error", "ComfyUI reported a workflow execution error.")
        if (clock() - started) >= IMAGE_TIMEOUT_SECONDS:
            return None, ProviderImageError("timeout", "Local image provider request timed out.")
        sleep(COMFYUI_POLL_INTERVAL_SECONDS)


def _first_image_ref(outputs: object) -> dict[str, str] | None:
    if not isinstance(outputs, dict):
        return None
    for node in outputs.values():
        images = node.get("images") if isinstance(node, dict) else None
        if not isinstance(images, list):
            continue
        for item in images:
            filename = item.get("filename") if isinstance(item, dict) else None
            if isinstance(filename, str) and filename:
                subfolder = item.get("subfolder")
                kind = item.get("type")
                return {
                    "filename": filename,
                    "subfolder": subfolder if isinstance(subfolder, str) else "",
                    "type": kind if isinstance(kind, str) else "output",
                }
    return None


def _fetch_comfyui_image(
    base: str,
    ref: dict[str, str],
    opener: Callable[..., Any] | None,
) -> tuple[bytes | None, ProviderImageError | None]:
    query = urlencode({"filename": ref["filename"], "subfolder": ref["subfolder"], "type": ref["type"]})
    url = urljoin(base + "/", "view") + "?" + query
    try:
        raw = get_bytes(url, timeout=IMAGE_TIMEOUT_SECONDS, opener=opener, max_bytes=IMAGE_MAX_BYTES)
    except HTTPProviderError as exc:
        if exc.code == "response_too_large":
            return None, ProviderImageError("image_too_large", "Local image exceeded the sandbox size limit.")
        return None, ProviderImageError(exc.code, _redact(exc.code))
    except Exception:  # noqa: BLE001 - never surface provider internals
        return None, ProviderImageError("provider_error", "Local image provider request failed.")
    return raw, None


# --- shared helpers --------------------------------------------------------------


def _validate_png(raw: bytes) -> ProviderImageError | None:
    if len(raw) > IMAGE_MAX_BYTES:
        return ProviderImageError("image_too_large", "Local image exceeded the sandbox size limit.")
    if not raw.startswith(PNG_SIGNATURE):
        return ProviderImageError("invalid_response", "Local image provider returned a non-PNG image.")
    return None


def _ok(
    started: float,
    clock: Callable[[], float],
    input_length: int,
    b64: str,
    image_bytes: int,
) -> ProviderImageReport:
    return ProviderImageReport(
        ok=True,
        attempted=True,
        configured=True,
        status="pass",
        input_length=input_length,
        duration_ms=_ms(started, clock),
        image_base64=b64,
        mime="image/png",
        image_bytes=image_bytes,
        untrusted_output=True,
        error=None,
    )


def _validate_prompt(prompt: str) -> tuple[str | None, ProviderImageError | None]:
    if len(prompt) > IMAGE_MAX_PROMPT_CHARS:
        return None, ProviderImageError("input_too_long", f"Image prompt exceeds {IMAGE_MAX_PROMPT_CHARS} characters.")
    if any(unicodedata.category(c) in {"Cc", "Cf", "Cs"} and c not in "\n\r\t" for c in prompt):
        return None, ProviderImageError("invalid_input", "Image prompt contains unsupported control characters.")
    stripped = prompt.strip()
    if not stripped:
        return None, ProviderImageError("empty_input", "Image prompt must not be empty.")
    return stripped, None


def _redact(code: str) -> str:
    return {
        "not_configured": "Private local image provider is not configured.",
        "connection_failed": "Local image provider connection failed.",
        "timeout": "Local image provider request timed out.",
        "http_error": "Local image provider returned an HTTP error.",
        "invalid_json": "Local image provider returned invalid JSON.",
        "response_too_large": "Local image provider response exceeded the sandbox limit.",
    }.get(code, "Local image provider request failed.")


def _ms(started: float, clock: Callable[[], float]) -> int:
    return max(0, round((clock() - started) * 1000))


def _fail(
    started: float,
    clock: Callable[[], float],
    code: str,
    message: str,
    *,
    status: str = "fail",
    configured: bool = False,
    input_length: int = 0,
    attempted: bool = False,
) -> ProviderImageReport:
    return ProviderImageReport(
        ok=False,
        attempted=attempted,
        configured=configured,
        status=status,
        input_length=input_length,
        duration_ms=_ms(started, clock),
        image_base64=None,
        mime=None,
        image_bytes=0,
        untrusted_output=True,
        error=ProviderImageError(code=code, message=message),
    )

File: workbench/src-tauri/src/bridge/provider_image_gen.rs
//! Approval-gated local image generation bridge.
//!
//! The frontend submits only a bounded prompt and acknowledgement boolean. Rust
//! owns the executable and argv; the prompt is written only to child stdin. The
//! backend (ComfyUI or an OpenAI-compatible image server) is chosen entirely by
//! the user's gitignored home config — never by the frontend. Output is one
//! sanitized base64 PNG, marked LOCAL UNTRUSTED; no model, key, path, or workflow
//! is ever returned.

use super::resolve_python::resolve_python;
use super::types::BridgeError;
use super::workspace::{get_workspace_resolution, WorkspaceResolutionStatus};
use serde::{Deserialize, Serialize};
use std::io::{Read, Write};
use std::path::{Path, PathBuf};
use std::process::{Command, ExitStatus, Stdio};
use std::thread;
use std::time::{Duration, Instant};

pub const IMAGE_MAX_PROMPT_CHARS: usize = 2_000;
pub const IMAGE_MAX_PROMPT_BYTES: usize = 8 * 1024;
pub const IMAGE_TIMEOUT_MS: u64 = 90_000; // > Python's 60s image timeout
pub const IMAGE_MAX_STDOUT_BYTES: usize = 16 * 1024 * 1024; // base64 PNG (~12 MB) + JSON
pub const IMAGE_MAX_STDERR_BYTES: usize = 8 * 1024;
pub const IMAGE_PYTHON_MODULE: &str = "realforge.cli";
pub const IMAGE_ARGV: &[&str] = &["provider", "image", "--stdin", "--json"];

const PASSTHROUGH_ENV: &[&str] = &[
    "PATH",
    "HOME",
    "USERPROFILE",
    "SYSTEMROOT",
    "WINDIR",
    "TEMP",
    "TMP",
];

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ProviderImageGenInput {
    pub prompt: String,
    pub approval_acknowledged: bool,
}

#[derive(Debug, Clone, Deserialize)]
struct CliImageError {
    code: String,
    #[allow(dead_code)]
    message: String,
}

#[derive(Debug, Clone, Deserialize)]
struct CliImageReport {
    ok: bool,
    attempted: bool,
    configured: bool,
    status: String,
    input_length: usize,
    duration_ms: u64,
    image_base64: Option<String>,
    mime: Option<String>,
    image_bytes: usize,
    #[allow(dead_code)]
    untrusted_output: bool,
    error: Option<CliImageError>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub struct ProviderImageError {
    pub code: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct ProviderImageReport {
    pub ok: bool,
    pub attempted: bool,
    pub configured: bool,
    pub status: String,
    pub input_length: usize,
    pub duration_ms: u64,
    pub image_base64: Option<String>,
    pub mime: Option<String>,
    pub image_bytes: usize,
    pub untrusted_output: bool,
    pub error: Option<ProviderImageError>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderImageGenResult {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<ProviderImageReport>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<BridgeError>,
}

impl ProviderImageGenResult {
    fn success(data: ProviderImageReport) -> Self {
        Self {
            ok: true,
            data: Some(data),
            error: None,
        }
    }

    pub(crate) fn failure(code: &str, message: impl Into<String>) -> Self {
        Self {
            ok: false,
            data: None,
            error: Some(BridgeError {
                code: code.to_string(),
                message: message.into(),
            }),
        }
    }
}

// ponytail: no single-flight guard/cancel for v1 — the UI disables Generate while a
// request runs and this is a single-user desktop app. Add the chat-style guard if
// concurrent image jobs ever become a real problem.
pub fn run_private_provider_image_gen(input: ProviderImageGenInput) -> ProviderImageGenResult {
    if !input.approval_acknowledged {
        return ProviderImageGenResult::failure(
            "approval_required",
            "Explicit approval is required before generating an image.",
        );
    }
    let prompt = match validate_prompt(&input.prompt) {
        Ok(prompt) => prompt,
        Err(error) => return ProviderImageGenResult::failure(&error.code, error.message),
    };

    let resolution = get_workspace_resolution();
    if resolution.status != WorkspaceResolutionStatus::Ready {
        return ProviderImageGenResult::failure(
            "workspace_not_ready",
            "The RealForge workspace and Python runtime must be ready before generating images.",
        );
    }
    let Some(repo_root) = resolution.repo_root.map(PathBuf::from) else {
        return ProviderImageGenResult::failure(
            "workspace_not_ready",
            "The RealForge workspace is unavailable.",
        );
    };
    let python = match resolve_python(&repo_root) {
        Ok(path) => path,
        Err(_) => {
            return ProviderImageGenResult::failure(
                "executable_not_found",
                "The configured RealForge Python runtime is unavailable.",
            )
        }
    };
    if python.components().count() > 1 && !python.is_file() {
        return ProviderImageGenResult::failure(
            "executable_not_found",
            "The configured RealForge Python runtime is unavailable.",
        );
    }

    match run_image_command(&python, &repo_root, &prompt) {
        Ok(report) => ProviderImageGenResult::success(report),
        Err(error) => ProviderImageGenResult::failure(&error.code, error.message),
    }
}

fn validate_prompt(prompt: &str) -> Result<String, BridgeError> {
    if prompt.chars().count() > IMAGE_MAX_PROMPT_CHARS {
        return Err(BridgeError {
            code: "input_too_long".to_string(),
            message: format!("Image prompt exceeds {IMAGE_MAX_PROMPT_CHARS} characters."),
        });
    }
    if prompt.len() > IMAGE_MAX_PROMPT_BYTES {
        return Err(BridgeError {
            code: "input_too_large".to_string(),
            message: "Image prompt exceeds the byte limit.".to_string(),
        });
    }
    if prompt
        .chars()
        .any(|character| character.is_control() && !matches!(character, '\n' | '\r' | '\t'))
    {
        return Err(BridgeError {
            code: "invalid_input".to_string(),
            message: "Image prompt contains unsupported control characters.".to_string(),
        });
    }
    let trimmed = prompt.trim();
    if trimmed.is_empty() {
        return Err(BridgeError {
            code: "empty_input".to_string(),
            message: "Image prompt must not be empty.".to_string(),
        });
    }
    Ok(trimmed.to_string())
}

fn run_image_command(
    python: &Path,
    repo_root: &Path,
    prompt: &str,
) -> Result<ProviderImageReport, BridgeError> {
    let mut command = Command::new(python);
    command
        .env_clear()
        .arg("-m")
        .arg(IMAGE_PYTHON_MODULE)
        .args(IMAGE_ARGV)
        .current_dir(repo_root)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .env("PYTHONPATH", repo_root.join("src"))
        .env("PYTHONDONTWRITEBYTECODE", "1")
        .env("PYTHONNOUSERSITE", "1")
        .env("PYTHONUTF8", "1")
        .env("LANG", "C.UTF-8");
    for name in PASSTHROUGH_ENV {
        if let Some(value) = std::env::var_os(name) {
            command.env(name, value);
        }
    }

    let mut child = command.spawn().map_err(|_| BridgeError {
        code: "spawn_failed".to_string(),
        message: "Could not start the image generation command.".to_string(),
    })?;
    let Some(mut stdin) = child.stdin.take() else {
        terminate_child(&mut child);
        return Err(BridgeError {
            code: "spawn_failed".to_string(),
            message: "Image generation stdin was unavailable.".to_string(),
        });
    };
    if stdin.write_all(prompt.as_bytes()).is_err() {
        terminate_child(&mut child);
        return Err(BridgeError {
            code: "input_failed".to_string(),
            message: "Image prompt could not be delivered.".to_string(),
        });
    }
    drop(stdin);

    let Some(stdout) = child.stdout.take() else {
        terminate_child(&mut child);
        return Err(BridgeError {
            code: "spawn_failed".to_string(),
            message: "Image generation stdout was unavailable.".to_string(),
        });
    };
    let Some(stderr) = child.stderr.take() else {
        terminate_child(&mut child);
        return Err(BridgeError {
            code: "spawn_failed".to_string(),
            message: "Image generation stderr was unavailable.".to_string(),
        });
    };
    let stdout_reader = thread::spawn(move || read_capped(stdout, IMAGE_MAX_STDOUT_BYTES));
    let stderr_reader = thread::spawn(move || read_capped(stderr, IMAGE_MAX_STDERR_BYTES));
    let timeout = Duration::from_millis(IMAGE_TIMEOUT_MS);
    let started = Instant::now();

    let status = loop {
        match child.try_wait() {
            Ok(Some(status)) => break status,
            Ok(None) if started.elapsed() >= timeout => {
                terminate_child(&mut child);
                let _ = stdout_reader.join();
                let _ = stderr_reader.join();
                return Err(BridgeError {
                    code: "timeout".to_string(),
                    message: "Image generation timed out before returning a result.".to_string(),
                });
            }
            Ok(None) => thread::sleep(Duration::from_millis(50)),
            Err(_) => {
                terminate_child(&mut child);
                let _ = stdout_reader.join();
                let _ = stderr_reader.join();
                return Err(BridgeError {
                    code: "spawn_failed".to_string(),
                    message: "Image generation process monitoring failed.".to_string(),
                });
            }
        }
    };

    let stdout = join_capped_output(stdout_reader, "stdout", IMAGE_MAX_STDOUT_BYTES)?;
    let _stderr = join_capped_output(stderr_reader, "stderr", IMAGE_MAX_STDERR_BYTES)?;
    parse_image_output(&stdout, status)
}

fn terminate_child(child: &mut std::process::Child) {
    let _ = child.kill();
    let _ = child.wait();
}

fn parse_image_output(
    stdout: &str,
    status: ExitStatus,
) -> Result<ProviderImageReport, BridgeError> {
    let parsed = serde_json::from_str::<CliImageReport>(stdout).map_err(|_| BridgeError {
        code: if status.success() {
            "invalid_json".to_string()
        } else {
            "image_failed".to_string()
        },
        message: "Image generation did not return the expected sanitized JSON report."
            .to_string(),
    })?;
    Ok(sanitize_report(parsed))
}

fn sanitize_report(report: CliImageReport) -> ProviderImageReport {
    let status = match report.status.as_str() {
        "pass" => "pass",
        "not_configured" => "not_configured",
        "rejected" => "rejected",
        _ => "fail",
    };
    // Only forward image bytes on a real pass; never trust a non-PNG mime.
    let mime = match report.mime.as_deref() {
        Some("image/png") if status == "pass" => Some("image/png".to_string()),
        _ => None,
    };
    let image_base64 = if status == "pass" { report.image_base64 } else { None };
    let image_bytes = if image_base64.is_some() {
        report.image_bytes
    } else {
        0
    };
    ProviderImageReport {
        ok: report.ok && status == "pass",
        attempted: report.attempted,
        configured: report.configured,
        status: status.to_string(),
        input_length: report.input_length.min(IMAGE_MAX_PROMPT_CHARS),
        duration_ms: report.duration_ms.min(IMAGE_TIMEOUT_MS),
        image_base64,
        mime,
        image_bytes,
        untrusted_output: true,
        error: sanitize_error(report.error, status),
    }
}

fn sanitize_error(error: Option<CliImageError>, status: &str) -> Option<ProviderImageError> {
    if status == "pass" {
        return None;
    }
    let code = error
        .as_ref()
        .map(|value| value.code.as_str())
        .unwrap_or("provider_error");
    let (code, message) = match code {
        "empty_input" => ("empty_input", "Image prompt must not be empty."),
        "input_too_long" => ("input_too_long", "Image prompt exceeds the character limit."),
        "input_too_large" => ("input_too_large", "Image prompt exceeds the byte limit."),
        "invalid_input" => (
            "invalid_input",
            "Image prompt contains unsupported characters.",
        ),
        "not_configured" => (
            "not_configured",
            "Private local image provider is not configured.",
        ),
        "invalid_config" => (
            "invalid_config",
            "Private local provider configuration is invalid.",
        ),
        "connection_failed" => ("connection_failed", "Local image provider connection failed."),
        "timeout" => ("timeout", "Local image provider request timed out."),
        "http_error" => ("http_error", "Local image provider returned an HTTP error."),
        "invalid_json" => ("invalid_json", "Local image provider returned invalid JSON."),
        "invalid_response" => (
            "invalid_response",
            "Local image provider returned an unsupported image.",
        ),
        "image_too_large" => (
            "image_too_large",
            "Local image exceeded the sandbox size limit.",
        ),
        "response_too_large" => (
            "response_too_large",
            "Local image provider response exceeded the sandbox limit.",
        ),
        "workflow_no_placeholder" => (
            "workflow_no_placeholder",
            "The ComfyUI workflow is missing the %prompt% token.",
        ),
        "invalid_workflow" => ("invalid_workflow", "The ComfyUI workflow is not valid JSON."),
        "workflow_missing" => (
            "workflow_missing",
            "The configured ComfyUI workflow file could not be read.",
        ),
        "workflow_too_large" => (
            "workflow_too_large",
            "The configured ComfyUI workflow file is too large.",
        ),
        _ => ("provider_error", "Local image provider request failed."),
    };
    Some(ProviderImageError {
        code: code.to_string(),
        message: message.to_string(),
    })
}

#[derive(Debug)]
struct CappedOutput {
    bytes: Vec<u8>,
    exceeded: bool,
}

fn read_capped(mut reader: impl Read, max_bytes: usize) -> Result<CappedOutput, std::io::Error> {
    let mut bytes = Vec::new();
    let mut exceeded = false;
    let mut buffer = [0u8; 8192];
    loop {
        let read = reader.read(&mut buffer)?;
        if read == 0 {
            break;
        }
        let remaining = max_bytes.saturating_sub(bytes.len());
        let retained = read.min(remaining);
        bytes.extend_from_slice(&buffer[..retained]);
        exceeded |= retained < read;
    }
    Ok(CappedOutput { bytes, exceeded })
}

fn join_capped_output(
    reader: thread::JoinHandle<Result<CappedOutput, std::io::Error>>,
    stream: &str,
    max_bytes: usize,
) -> Result<String, BridgeError> {
    let output = reader
        .join()
        .map_err(|_| BridgeError {
            code: "read_failed".to_string(),
            message: format!("Image generation {stream} reader failed."),
        })?
        .map_err(|_| BridgeError {
            code: "read_failed".to_string(),
            message: format!("Image generation {stream} could not be read."),
        })?;
    if output.exceeded {
        return Err(BridgeError {
            code: "output_too_large".to_string(),
            message: format!("Image generation {stream} exceeded the {max_bytes}-byte limit."),
        });
    }
    String::from_utf8(output.bytes).map_err(|_| BridgeError {
        code: "invalid_output".to_string(),
        message: format!("Image generation {stream} was not valid UTF-8."),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn cli_report() -> CliImageReport {
        CliImageReport {
            ok: true,
            attempted: true,
            configured: true,
            status: "pass".to_string(),
            input_length: 5,
            duration_ms: 42,
            image_base64: Some("iVBORw0KGgo=".to_string()),
            mime: Some("image/png".to_string()),
            image_bytes: 68,
            untrusted_output: false,
            error: None,
        }
    }

    #[test]
    fn input_accepts_only_prompt_and_approval() {
        let parsed = serde_json::from_str::<ProviderImageGenInput>(
            r#"{"prompt":"a fox","approvalAcknowledged":true}"#,
        )
        .unwrap();
        assert_eq!(parsed.prompt, "a fox");
        assert!(parsed.approval_acknowledged);
        for extra in ["args", "path", "model", "endpoint", "workflow"] {
            let payload =
                format!(r#"{{"prompt":"x","approvalAcknowledged":true,"{extra}":"blocked"}}"#);
            assert!(serde_json::from_str::<ProviderImageGenInput>(&payload).is_err());
        }
    }

    #[test]
    fn command_shape_is_fixed_and_stdin_only() {
        assert_eq!(IMAGE_PYTHON_MODULE, "realforge.cli");
        assert_eq!(IMAGE_ARGV, &["provider", "image", "--stdin", "--json"]);
    }

    #[test]
    fn prompt_validation_is_bounded() {
        assert_eq!(validate_prompt("  a cat  ").unwrap(), "a cat");
        assert_eq!(validate_prompt("   ").unwrap_err().code, "empty_input");
        assert_eq!(
            validate_prompt(&"X".repeat(IMAGE_MAX_PROMPT_CHARS + 1))
                .unwrap_err()
                .code,
            "input_too_long"
        );
        assert_eq!(
            validate_prompt("bad\0prompt").unwrap_err().code,
            "invalid_input"
        );
    }

    #[test]
    fn sanitizer_forces_untrusted_and_keeps_png_on_pass() {
        let report = sanitize_report(cli_report());
        assert!(report.untrusted_output);
        assert!(report.ok);
        assert_eq!(report.mime.as_deref(), Some("image/png"));
        assert!(report.image_base64.is_some());
    }

    #[test]
    fn sanitizer_drops_image_and_redacts_error_on_failure() {
        let mut source = cli_report();
        source.ok = false;
        source.status = "fail".to_string();
        source.mime = Some("image/png".to_string());
        source.image_base64 = Some("leaked".to_string());
        source.error = Some(CliImageError {
            code: "unknown-private-code".to_string(),
            message: "private transport detail".to_string(),
        });
        let report = sanitize_report(source);
        let serialized = serde_json::to_string(&report).unwrap();
        assert_eq!(report.image_base64, None);
        assert_eq!(report.mime, None);
        assert_eq!(report.image_bytes, 0);
        assert!(!serialized.contains("leaked"));
        assert!(!serialized.contains("private transport detail"));
        assert_eq!(report.error.unwrap().code, "provider_error");
    }

    #[test]
    fn sanitizer_maps_comfyui_workflow_error() {
        let mut source = cli_report();
        source.ok = false;
        source.status = "fail".to_string();
        source.error = Some(CliImageError {
            code: "workflow_no_placeholder".to_string(),
            message: "detail".to_string(),
        });
        let report = sanitize_report(source);
        assert_eq!(report.error.unwrap().code, "workflow_no_placeholder");
    }

    #[test]
    fn serialized_report_has_no_private_fields() {
        let report = sanitize_report(cli_report());
        let serialized = serde_json::to_value(report).unwrap();
        let object = serialized.as_object().unwrap();
        assert!(!object.contains_key("model"));
        assert!(!object.contains_key("api_key"));
        assert!(!object.contains_key("workflow"));
        assert!(!object.contains_key("endpoint"));
        assert!(!object.contains_key("prompt"));
    }
}
